### PR TITLE
Auditability

### PR DESCRIPTION
**Pull request by dholms**
Adds two new routes
- `/export` which exports the _entire_ PLC database as a series of jsonlines, where each line is an operation ordered by `createdAt` time. This is paginated with a limit of 1000 lines per req
- `/:did/auditable` which gives the auditable log for a user that includes timestamps & nullified operations

### CODE REVIEW

File: packages/lib/tests/compatibility.test.ts [hunk 1]
```diff
@@ -78,6 +78,6 @@ describe('compatibility', () => {
 
     const result = await assureValidNextOp(did, [indexedLegacy], nextOp)
     expect(result.nullified.length).toBe(0)
-    expect(result.prev?.equals(legacyCid))
+    expect(result.prev?.equals(legacyCid)).toBeTruthy()
```
Comment (devinivy) [inline packages/lib/tests/compatibility.test.ts:81]: 👀

👍 nice catch

File: packages/server/src/routes.ts [hunk 2]
```diff
@@ -18,15 +18,20 @@ export const createRouter = (ctx: AppContext): express.Router => {
     res.send({ version })
   })
 
-  // @TODO paginate & test this
+  // Export ops in the form of paginated json lines
   router.get('/export', async function (req, res) {
-    const fullExport = await ctx.db.fullExport()
+    const parsedCount = parseInt(req.query.count)
+    const count = isNaN(parsedCount) ? 1000 : Math.min(parsedCount, 1000)
```
Comment (devinivy) [inline packages/server/src/routes.ts:24]: To help avoid accepting weird stuff including numbers written in hex for the `count`:
```suggestion
    const { count } = req.query
    const parsedCount = count ? parseInt(count, 10) : 1000
    if (isNaN(parsedCount) || parsedCount < 1) {
      throw new ServerError(400, 'Invalid count parameter')
    }
    const count = Math.min(parsedCount, 1000)
```

File: packages/server/src/routes.ts [hunk 3]
```diff
@@ -48,7 +53,7 @@ export const createRouter = (ctx: AppContext): express.Router => {
   })
 
   // Get data for a DID document
-  router.get('/data/:did', async function (req, res) {
+  router.get('/:did/data', async function (req, res) {
```
Comment (devinivy) [inline packages/server/src/routes.ts:56]: 👍 Big +1 on these changes!

File: packages/server/src/routes.ts [hunk 4]
```diff
@@ -62,24 +67,39 @@ export const createRouter = (ctx: AppContext): express.Router => {
...
+      cid: op.cid.toString(),
+      createdAt: op.createdAt.toISOString(),
+    }))
+
+    res.json(log)
   })
 
   // Get the most recent operation in the log for a DID
-  router.get('/last/:did', async function (req, res) {
+  router.get('/:did/last', async function (req, res) {
```
Comment (devinivy) [inline packages/server/src/routes.ts:96]: In a similar line of thought re: `/:did/log/audit`, I could also imagine calling this `/:did/log/last`.  Really not too hung-up on it, really just one final poke to make sure we like the naming conventions here!

**Review by devinivy — APPROVED** [review]
PLC just got a little comfier and cozier 👍

File: packages/server/src/routes.ts (hunk 2 above)
Comment (dholms) [reply packages/server/src/routes.ts:24]: nice i like it 👍

File: packages/server/src/routes.ts (hunk 3 above)
Comment (dholms) [reply packages/server/src/routes.ts:60]: yuppp

File: packages/server/src/routes.ts [hunk 5]
```diff
@@ -62,24 +67,39 @@ export const createRouter = (ctx: AppContext): express.Router => {
...
     const log = await ctx.db.opsForDid(did)
     if (log.length === 0) {
       throw new ServerError(404, `DID not registered: ${did}`)
     }
-    res.json({ log })
+    res.json(log)
+  })
+
+  // Get operation log for a DID
+  router.get('/:did/auditable', async function (req, res) {
```
[the remark this answers is not in the document]
Comment (dholms) [reply packages/server/src/routes.ts:84]: good point

honestly maybe just `/:did/audit`

File: packages/server/src/routes.ts (hunk 4 above)
Comment (dholms) [reply packages/server/src/routes.ts:100]: actually you're right i'll do those 👍

**Pull request merged**